Add helm installation steps for multi-cluster (#15459)
istio/istio.io: content/en/docs/setup/install/multicluster/primary-remote/index.md

[appropriately annotated](#set-the-control-plane-cluster-for-cluster2) remote clusters that will be
attached to it (`cluster2` in this case).

{{< /tab >}}

{{< tab name="Helm" category-value="helm" >}}

> Install Istio as primary in `cluster1` using standard `helm` commands.

Reviewer: ```suggestion
Install Istio as primary in `cluster1` using standard Helm commands.
```

First, install the `base` chart in `cluster1`:

{{< text bash >}}

> $ helm install istio-base istio/base -n istio-system --set global.externalIstiod=true --kube-context "${CTX_CLUSTER1}"

Reviewer 2: FWIW, `externalIstiod` is not used or templated anywhere in the `base` chart that I can see - it does not need to be set for `base` at all.

If the control-plane was installed with a revision, add the `--revision rev` flag to the `gen-eastwest-gateway.sh` command.
{{< /warning >}}

{{< /tab >}}
{{< tab name="Helm" category-value="helm" >}}

> Install the east-west gateway in `cluster1` using standard `helm` commands:

Reviewer: ```suggestion
Install the east-west gateway in `cluster1` using standard Helm commands:
```

{{< text bash >}}
$ helm install istio-eastwestgateway istio/gateway -n istio-system --kube-context "${CTX_CLUSTER1}" --set name=istio-eastwestgateway --set networkGateway=network1
{{< /text >}}

{{< warning >}}

> If the control-plane was installed with a revision, add the `--set revision=rev` flag to the `helm` install command.

Reviewer: ```suggestion
If the control-plane was installed with a revision, add the `--set revision=rev` flag to the Helm install command.
```

> If the control-plane was installed with a revision, add the `--set revision=rev` flag to the Helm install command.

Reviewer 2: ```suggestion
If the control-plane was installed with a revision, you must add a `--set revision=<my-revision>` flag to the Helm install command.
```

EOF
{{< /text >}}

Apply the configuration to `cluster2`:

{{< text bash >}}
$ istioctl install --context="${CTX_CLUSTER2}" -f cluster2.yaml
{{< /text >}}

{{< /tab >}}
{{< tab name="Helm" category-value="helm" >}}

Install the `istiod-remote` chart in `cluster2` using standard `helm` commands:

{{< text bash >}}

> helm install istiod-remote istio/istiod-remote --set global.multiCluster.clusterName=cluster2 --set global.network=network1 --set istiodRemote.injectionPath=/inject/cluster/cluster2/net/network1 --set global.externalIstiod=true --set global.configCluster=true --set global.remotePilotAddress=${DISCOVERY_ADDRESS} --set pilot.enabled=false -n istio-system --kube-context ${CTX_CLUSTER2}

Author: @bleggett @howardjohn one of the improvements we could potentially make would be to consolidate some of these vars in the remote chart - ex: configCluster and pilot.enabled maybe?

Author: Should we still publish these steps for now and update the doc as a follow-up later, or wait until the `istiod-remote` chart has been updated with the new values before publishing in the first place? Not sure if that will make it to `1.23`.

Reviewer 2: I'd be inclined to document actual/current state and fix it later.

EOF
{{< /text >}}

Apply the configuration to `cluster2`:

{{< text bash >}}
$ istioctl install --context="${CTX_CLUSTER2}" -f cluster2.yaml
{{< /text >}}

{{< /tab >}}
{{< tab name="Helm" category-value="helm" >}}

> Install the `istiod-remote` chart in `cluster2` using standard `helm` commands:

Reviewer: ```suggestion
Install the `istiod-remote` chart in `cluster2` using standard Helm commands:
```

{{< text bash >}}
$ helm install istiod-remote istio/istiod-remote --set global.multiCluster.clusterName=cluster2 --set global.network=network1 --set istiodRemote.injectionPath=/inject/cluster/cluster2/net/network1 --set global.configCluster=true --set global.remotePilotAddress="${DISCOVERY_ADDRESS}" --set pilot.enabled=false -n istio-system --kube-context "${CTX_CLUSTER2}"
{{< /text >}}

{{< warning >}}

> The `istiod-remote` Helm chart is only available from Istio release `1.23` onwards.

Reviewer: ```suggestion
The `istiod-remote` Helm chart is only available from Istio release 1.23 onwards.
```

EOF
{{< /text >}}

Apply the configuration to `cluster2`:

{{< text bash >}}
$ istioctl install --context="${CTX_CLUSTER2}" -f cluster2.yaml
{{< /text >}}

{{< /tab >}}
{{< tab name="Helm" category-value="helm" >}}

> Install Istio as remote in `cluster2` using standard Helm commands.

Reviewer 2: ```suggestion
Install Istio as remote in `cluster2` using the following Helm commands:
```

> First, install the `base` chart in `cluster2`:

Reviewer 2: ```suggestion
Install the `base` chart in `cluster2`:
```

{{< text bash >}}
$ helm install istio-base istio/base --version 1.24.0-alpha.0 -n istio-system --set profile=remote --set global.externalIstiod=true --kube-context "${CTX_CLUSTER2}"
{{< /text >}}

Then, install the `istiod` chart in `cluster2` with the following multi-cluster settings:

{{< text bash >}}
$ helm install istiod istio/istiod --version 1.24.0-alpha.0 -n istio-system --set profile=remote --set global.externalIstiod=true --set pilot.env.EXTERNAL_ISTIOD=true --set global.multiCluster.clusterName=cluster2 --set istiodRemote.injectionPath=/inject/cluster/cluster2/net/network1 --set global.configCluster=true --set global.remotePilotAddress="${DISCOVERY_ADDRESS}" --kube-context "${CTX_CLUSTER2}"
{{< /text >}}

> {{< warning >}}

Reviewer 2: Rather than a warning, make this a `note` or `suggestion`

## Cleanup

> Uninstall Istio from both `cluster1` and `cluster2` using `istioctl` or `helm`.

Reviewer: ```suggestion
Uninstall Istio from both `cluster1` and `cluster2` using istioctl or Helm.
```

{{< tabset category-name="multicluster-uninstall-type-cluster-1" >}}

{{< tab name="IstioOperator" category-value="iop" >}}

1. Uninstall Istio in `cluster1`:

> {{< text syntax=bash snip_id=none >}}
> $ istioctl uninstall --context="${CTX_CLUSTER1}" -y --purge
> $ kubectl delete ns istio-system --context="${CTX_CLUSTER1}"
> {{< /text >}}

Reviewer: ```suggestion
    {{< text syntax=bash snip_id=none >}}
    $ istioctl uninstall --context="${CTX_CLUSTER1}" -y --purge
    $ kubectl delete ns istio-system --context="${CTX_CLUSTER1}"
    {{< /text >}}
```

1. Uninstall Istio in `cluster2`:

> {{< text syntax=bash snip_id=none >}}
> $ istioctl uninstall --context="${CTX_CLUSTER2}" -y --purge
> $ kubectl delete ns istio-system --context="${CTX_CLUSTER2}"
> {{< /text >}}

Reviewer: ```suggestion
    {{< text syntax=bash snip_id=none >}}
    $ istioctl uninstall --context="${CTX_CLUSTER2}" -y --purge
    $ kubectl delete ns istio-system --context="${CTX_CLUSTER2}"
    {{< /text >}}
```

{{< /tab >}}

{{< tab name="Helm" category-value="helm" >}}

> 1. Delete Istio helm charts from `cluster1`:

Reviewer: ```suggestion
1. Delete Istio Helm installation from `cluster1`:
```

> {{< text syntax=bash >}}
> $ helm delete istiod -n istio-system --kube-context "${CTX_CLUSTER1}"
> $ helm delete istio-eastwestgateway -n istio-system --kube-context "${CTX_CLUSTER1}"
> $ helm delete istio-base -n istio-system --kube-context "${CTX_CLUSTER1}"
> {{< /text >}}

Reviewer: ```suggestion
    {{< text syntax=bash >}}
    $ helm delete istiod -n istio-system --kube-context "${CTX_CLUSTER1}"
    $ helm delete istio-eastwestgateway -n istio-system --kube-context "${CTX_CLUSTER1}"
    $ helm delete istio-base -n istio-system --kube-context "${CTX_CLUSTER1}"
    {{< /text >}}
```

1. Delete the `istio-system` namespace from `cluster1`:

> {{< text syntax=bash >}}
> $ kubectl delete ns istio-system --context="${CTX_CLUSTER1}"
> {{< /text >}}

Reviewer: ```suggestion
    {{< text syntax=bash >}}
    $ kubectl delete ns istio-system --context="${CTX_CLUSTER1}"
    {{< /text >}}
```

> 1. Delete Istio helm charts from `cluster2`:

Reviewer: ```suggestion
1. Delete Istio Helm installation from `cluster2`:
```

> {{< text syntax=bash >}}
> $ helm delete istiod-remote -n istio-system --kube-context "${CTX_CLUSTER2}"
> {{< /text >}}

Reviewer: ```suggestion
    {{< text syntax=bash >}}
    $ helm delete istiod-remote -n istio-system --kube-context "${CTX_CLUSTER2}"
    {{< /text >}}
```

1. Delete the `istio-system` namespace from `cluster2`:

> {{< text syntax=bash >}}
> $ kubectl delete ns istio-system --context="${CTX_CLUSTER2}"
> {{< /text >}}

Reviewer: ```suggestion
    {{< text syntax=bash >}}
    $ kubectl delete ns istio-system --context="${CTX_CLUSTER2}"
    {{< /text >}}
```

1. (Optional) Deleting CRDs installed by Istio

> Deleting CRDs permanently removes any Istio resources you have created in your clusters.
> To delete Istio CRDs installed in your clusters:
> 
> {{< text syntax=bash snip_id=delete_crds >}}
> $ kubectl get crd -oname --context "${CTX_CLUSTER1}" | grep --color=never 'istio.io' | xargs kubectl delete --context "${CTX_CLUSTER1}"
> $ kubectl get crd -oname --context "${CTX_CLUSTER2}" | grep --color=never 'istio.io' | xargs kubectl delete --context "${CTX_CLUSTER2}"
> {{< /text >}}

Reviewer: ```suggestion
    Deleting CRDs permanently removes any Istio resources you have created in your clusters.
    To delete Istio CRDs installed in your clusters:

    {{< text syntax=bash snip_id=delete_crds >}}
    $ kubectl get crd -oname --context "${CTX_CLUSTER1}" | grep --color=never 'istio.io' | xargs kubectl delete --context "${CTX_CLUSTER1}"
    $ kubectl get crd -oname --context "${CTX_CLUSTER2}" | grep --color=never 'istio.io' | xargs kubectl delete --context "${CTX_CLUSTER2}"
    {{< /text >}}
```

## Cleanup

> Uninstall Istio from both `cluster1` and `cluster2` using istioctl or Helm.

Reviewer 2: ```suggestion
Uninstall Istio from both `cluster1` and `cluster2` using the same mechanism you installed with (istioctl or Helm).
```
